feature: delete records

jniles opened the pull request:
This PR implements  deleting cash payments following the proposal in #2044 and the design discussed in #2159.  It does much of the heavy lifting for the other two delete issues - vouchers (#2161) and invoices (#2160).

TODO:
 - [x] End to End Tests
 - [x] Delete Invoices #2160.
 - [x] Delete Vouchers #2161.

jniles committed:
fix: spell referenceLookup correctly

This commit fixes a pervasive typo throughout the app where reference
was spelled `refence`.  This has been corrected.

jniles committed:
chore: prune dead cash route /references

This commit prunes the server to remove the cash route "/references"
that was unused.  If it ever found to be used in the future, this is the
place you want to restore it from.

jniles committed:
refactor: make errors/index.js

This commit makes an errors/index.js file to enable easy requiring of
multiple error handlers.  For example:
```js
const {
  BadRequest,
  NotFound
} = require('../lib/errors');
```

jniles committed:
feat(cash): delete cash payments

This commit implements the server-side logic and code for deleting cash
payments.  An integration tests shows that the functionality is intact.
This goes halfway to addressing #2159.

jniles committed:
feat(cash): implement client delete record

This commit implements the client-side for deleting a cash record.  It
simply adds a new value to the dropdown that can be clicked to remove
the record.

TODO - add end-to-end tests.

Closes #2159.

jniles committed:
fix: build issues from previous merge

This commit fixes improper SQL merge issues from previous merges.

jniles committed:
feat(invoices): implement delete server-side

This commit implements the DELETE /transactions/:uuid route for invoices
to remove records from the database.

Closes #2060.

jniles committed:
feat(vouchers): implement delete server-side

This commit implements the DELETE /transactions/:uuid route for the
vouchers route.

Closes #2161.

jniles committed:
chore(vouchesr): implement dropdown for delete

This commit implements the delete link on the Voucher Registry dropdown.

jniles committed:
chore: implement client side for deleting records

This commit polishes the client side code for deleting records from the
registries.  It also ensure the delete integration tests for invoices
reference the correct invoice.

Closes #2160.

jniles committed:
test: end-to-end record deletion

This commit implements end-to-end tests for record deletions.  It also
reworks the 'actions' menus on the cash and invoice registries to use a
similar layout, and builds the 'GridRow' shared class to harness those
menus.  This makes selection of each grid row explicit - the record
human readable reference is passed in to select it.

Closes #2044.

jniles committed:
fix: bowerrc parsing error

This commit fixes a wayword comma, causing bower to crash and Karma
tests not to run on our CI.

jniles committed:
feat: confirm before delete

This commit implements the ConfirmModal on the delete routes to make
sure that users do not haphazardly delete records.

DedrickEnc commented on client/src/i18n/fr/form.json:
```diff
@@ -66,6 +67,7 @@
         },
         "DIALOGS": {
             "CONFIRM_ACTION": "Confirmez vous cette action ?",
+            "CONFIRMATION_NEEDED" : "Confirmation Nécessaire",
```
Confirmation Requise is better

DedrickEnc commented on client/src/i18n/fr/form.json:
```diff
@@ -154,7 +156,8 @@
             "TRANSACTIONS": "Transactions",
-            "ROWS": "enregistrements"
+            "ROWS": "enregistrements",
+            "DELETE_RECORD_SUCCESS": "Supprimèe la transaction avec succès."
```
Suppression de la transaction avec succes or transaction supprimée avec succes.

DedrickEnc commented on client/src/modules/cash/payments/registry.js:
```diff
@@ -15,13 +15,12 @@ CashPaymentRegistryController.$inject = [
   var vm = this;
 
-  // Background color for make the difference between the valid and cancel paiement
+  // background color for make the difference between the valid and canceled payment
```
Good enhancement

DedrickEnc commented on client/src/modules/cash/payments/registry.js:
```diff
@@ -203,5 +203,26 @@ function CashPaymentRegistryController(
+      .then(function () {
+        Notify.success('FORM.INFO.DELETE_RECORD_SUCCESS');
+
+        // load() has it's own error handling.  The absence of return below is
```
Cool

DedrickEnc commented on client/src/modules/cash/payments/registry.js:
```diff
@@ -203,5 +203,26 @@ function CashPaymentRegistryController(
+
+  // this function deletes the cash payment and associated transactions from
+  // the database
+  function deleteCashPayment(entity) {
```
This function is asking for confirmation before removing a transaction, it should be named differently ```deleteCashConfirmation``` for example.

DedrickEnc commented on client/src/modules/invoices/registry/registry.js:
```diff
@@ -219,6 +221,26 @@ function InvoiceRegistryController(
+      .then(function () {
+        Notify.success('FORM.INFO.DELETE_RECORD_SUCCESS');
+
+        // load() has it's own error handling.  The absence of return below is
```
Good comment

DedrickEnc commented on client/src/modules/invoices/registry/registry.js:
```diff
@@ -219,6 +221,26 @@ function InvoiceRegistryController(
+  }
+
+  // check if it is okay to remove the entity.
+  function deleteInvoice(entity) {
```
A clear function name will be better

DedrickEnc replied:
A clear function name will be better

DedrickEnc commented on client/src/modules/vouchers/voucher-registry.ctrl.js:
```diff
@@ -233,5 +235,25 @@ function VoucherController(
+  }
+
+  // this function deletes the voucher from the database
+  function deleteVoucher(entity) {
```
Same comment

DedrickEnc replied:
Same comment

DedrickEnc commented on client/src/modules/vouchers/vouchers.service.js:
```diff
@@ -188,5 +190,18 @@ function VoucherService(Api, $http, TransactionTypeStore, Modal,
+   * This function removes a voucher from the database via the transaction
+   * delete route.
+   */
+  function remove(uuid) {
```
This code is seriously duplicated, no way to create a service to manage transaction as entity?

DedrickEnc replied:
@jniles 
This comment is not addressed.

DedrickEnc commented on server/controllers/finance/cash.js:
```diff
@@ -1,77 +1,54 @@
-function lookup(id) {
-  const bid = db.bid(id);
+function lookup(uuid) {
+  const bid = db.bid(uuid);
```
More clear +1

DedrickEnc commented on server/controllers/finance/cash.js:
```diff
@@ -122,26 +99,32 @@ function lookup(id) {
-  // ensure epected options are parsed appropriately as binary
-  db.convert(options, ['debtor_uuid', 'debtor_group_uuid']);
+function find(options) {
+  // ensure expected options are parsed appropriately as binary
```
+1

DedrickEnc commented on server/controllers/finance/cash.js:
```diff
@@ -1,77 +1,54 @@
-/** checkInvoicePayment if the invoice is paid */
+exports.find = find;
 exports.checkInvoicePayment = checkInvoicePayment;
+exports.safelyDeleteCashPayment = safelyDeleteCashPayment;
```
This function should not be exposed since it is deleting a record without checking the opportunity of doing that. It should be there for internal use.

jniles replied:
Where do you suggest it should go?

DedrickEnc replied:
Expose an other function which is doing checking first and then call the ```safelyDeleteCashPayment```.

DedrickEnc replied:
~What is about this issue ?~

DedrickEnc commented on server/controllers/finance/cash.js:
```diff
@@ -253,32 +235,11 @@ function update(req, res, next) {
-/**
  * GET /cash/:checkin/:invoiceUuid
  * Check if the invoice is paid
+ * TODO(@jniles) - this should use a more intelligent system to see if an
```
What will happen if the invoice is not paid but posted?

jniles replied:
~~It will be rejected by the `/transactions/:uuid` route.  Specifically, this line: https://github.com/IMA-WorldHealth/bhima-2.X/pull/2188/files#diff-3a4662a4da5c4801c73af27c7f7358ddR131~~

EDIT:  The above is not correct, I didn't see the associated comment - I thought this was on the function below.

At the moment, this only checks the cash table.  It would be best to check the balance of the invoice ... after all, invoices can also be paid by vouchers.

DedrickEnc replied:
Ha ha, it was strange because I saw nothing at this link also

DedrickEnc commented on test/integration/patientInvoice.js:
```diff
@@ -69,7 +68,7 @@ describe('(/invoices) Patient Invoices', function () {
       return agent.get('/invoices')
         .then((res) => {
-          helpers.api.listed(res, numInvoices + numCreatedInvoices);
+          helpers.api.listed(res, numInvoices + numCreatedInvoices - 1);
```
Why?

jniles replied:
One was deleted.  I supposed it could be in a removedInvoices variable.  I'll do that.

DedrickEnc commented on yarn.lock:
```diff
@@ -25,6 +25,10 @@ abbrev@1:
   version "1.1.0"
   resolved "https://registry.yarnpkg.com/abbrev/-/abbrev-1.1.0.tgz#d0554c2256636e2f56e7c2e5ad183f859428d81f"
 
+abbrev@^1.0.7:
```
This file is huge and you have say nothing about that in your PR description.

jniles replied:
It is generated automatically by yarn when you install/update dependencies.

DedrickEnc reviewed (commented):
@jniles

This P.R is huge, next time if a P.R will touch more that 20 files, we should create a feature branch to make a review step by step

This pull request deals with the invoice, cash, and voucher removing, it should be 3 different P.R

I have made some comments, just to be fixed about some dark areas.

Also, it seems like the test is not passing.

It is a huge work @jniles, thanks.

jniles committed:
feat: decouple transaction checks

This commit decouples the transaction checks from transaction controller
so that they can be shared between multiple controllers.  This ensures
safety in case a developers accidentally writes code that uses the
delete interface to delete transactions w/o ensuring the transactions
are properly dereferenced.
